pip/req/req_set.py

```diff
@@ -439,6 +461,11 @@ def _prepare_file(self, finder, req_to_install):
             # # vcs update or unpack archive # #
             # ################################ #
             if req_to_install.editable:
+                if require_hashes:
+                    raise InstallationError(
+                        'The editable requirement %s cannot be installed when '
+                        'requiring hashes, because there is no single file to '
+                        'hash.' % req_to_install)
```

Comment:
So yes, this is something I thought we might add in the future: pip would recognize hash-based git refspecs as okay for hash-checking mode. It would run `git fsck` over them to make sure the hashes really match. (git doesn't check them implicitly, at least in older versions.) My only reservation is that SHA1 isn't considered a very collision-resistant hash anymore.